### ISSUE TITLE
wave() function

### ISSUE THREAD

**Issue by AntikoreDev** [Alphish/gm-community-toolbox#21]
```gml
/// @func wave(from,to,duration,offset=0)
/// @desc Waves a value between two numbers with an specific duration.
/// @arg  {Real} The value to wave from.
/// @arg  {Real} The value to wave to.
/// @arg  {Real} How much time it will cost to do the animation.
/// @arg  {Real} Offset for the time.
/// @returns {Real} The value for the wave at that certain moment of time.

function wave(_from, _to, _duration, _offset = 0) {
	var _a4 = (_to - _from) * 0.5;
	return _from + a4 + sin((((current_time * 0.001) +  _duration * _offset) /  _duration) * (pi * 2)) *  _a4;
}
```
I always use this function for all my games, its just so simple and makes a nice effect on games.
It's an adaptation from one of the scripts shown in a Shaun Spalding video that I modified for modern GameMaker.

**Comment by dsk80004** [comment]
I know this is a common function and people are very use to the name `wave` but please lets not use single word function names, i have the same issue with the commonly used `approach`.

**Comment by FoxyOfJungle** [comment]
I never understood the real need for this function, if sin() already works for that (and is much faster). Unless you want it to move exactly in the given period in seconds.

**Comment by glebtsereteli** [comment]
> I know this is a common function and people are very use to the name `wave` but please lets not use single word function names, i have the same issue with the commonly used `approach`.

Agreed! If we were to add this, I'd suggest Mimpy's functions and (similar?) naming from this article: https://gm48.net/resource/33/sine-waves-make-your-game-prettier

**Comment by Alphish** [comment]
Yeah, `wave` is too general of a name, considering the varying implementations people might want to use (e.g. using midpoint + amplitude instead of top and bottom, using animation curves instead of sin etc.). Then there's also the matter of using exact time rather than letting the user pass in the current moment.

The linked `sine_wave` and `sine_between` functions would be better, even more so considering you can provide both the current time and the period (which allows the user to choose exact time or frames as a point of reference).

I'm still wondering about the exact naming. In particular, should we use word `sine` or more familiar `sin` (so `sin_between` and `sin_wave`), and what descriptors we want to use? E.g. `sine_wave` doesn't sound as descriptive as `sine_between` to me; maybe `sine_around` would be a better alternative...?

(as for using single-word function names - generally, I prefer not to overuse them either, but `approach` and other highly general math functions are an exception to me)

**Comment by tinkerer-red** [comment]
i think `sine_wave` makes sense personally, despite `sin` being the more common way to write it when using short hands, the phrase "sine wave" is not short hand and thus shouldnt use a shorthand world as a replacement.

I think `sine_wave` makes sense, arguably because of the sheer common hood of people using the `wave` naming format it's not really viable to remove the word "wave" from the function otherwise a ton of repos will have a duplicate of the exact same function.

**Comment by Mercerenies** [comment]
I'm not sure I see the benefit of this over just using `sin`. I suppose perhaps someone who hasn't had a trigonometry course would benefit from having the arguments spelled out, but it seems like this is just moving the goalpost from "understand `sin`" to "understand this transformed version of `sin`".

**Comment by AntikoreDev** [comment]
Yeah, probably `sine_wave` comes out as a better name.

> I'm not sure I see the benefit of this over just using `sin`. I suppose perhaps someone who hasn't had a trigonometry course would benefit from having the arguments spelled out, but it seems like this is just moving the goalpost from "understand `sin`" to "understand this transformed version of `sin`".

Not everyone has enough time or energy to understand maths completly, that's why this script comes out handy. Also, using only `sin` would probably fit for small things, but when you are doing this over and over throughout the whole game, having it on a separate script helps a lot too.

**Comment by Alphish** [comment]
Just for the record, are you people mentioning `sine_wave` as preferable to `sin_wave` specifically, or to `sine_around` as well?

Personally, I'm not familiar with using word `wave` to refer to the middle+amplitude set of parameters, as opposed to the `between` variant which uses min+max points instead. From that point of view, `sine_around` + `sine_between` to me conveys the relationship between the two more clearly than `sine_wave` + `sine_between`.

And I guess this discussion might become pretty relevant if we were to add similar functions for animation curves, for example. So e.g. there could be functions like `animcurve_cycle_around(curve,offset,period,midpoint,amplitude)` and `animcurve_cycle_between(curve,offset,period,min,max)`, or maybe the first would be called `animcurve_cycle_wave`? I don't know, but we might want to consider these conventions as well.

**Comment by AntikoreDev** [comment]
I just said `sine_wave` in general, I meant to say that probably any of those names you all said are better than just `wave` 😄

**Comment by gnysek** [comment]
I also think that we should not use functions whose names are a single word and such a noun/verb/adjective could be used by someone as a variable name (except pure math functions, but GM already covers most of them).
`sine_wave`, `sin_wave`, `get_wave`, `calc_wave` etc. are much better.

**Comment by Alphish** [comment]
Potential function name candidates: `cycle_around` and `cycle_between`. The former would use a given midpoint value and amplitude, while the latter would use a given minimum and maximum value.

By default, it would use a sine wave as a wave function, but it might be possible to override with an alternative function or an animation curve.
